### ISSUE TITLE
Feature Request: early stopping round tolerance

### ISSUE THREAD

**Issue by tecosaur** [Evovest/EvoTrees.jl#329]
Hello,

I was tuning a classifier tree for a problem, and I found that I was always hitting the maximum number of rounds, even though for the latter half the improvement was negligible.

Looking at the per-iteration performance, there's a point where iterations don't meaningfully improve, but the strict inequality here:

https://github.com/Evovest/EvoTrees.jl/blob/1e6c7a79b8d18124545f20737f1a0fef3fddf03d/src/callback.jl#L148-L149

means that an insignificant improvement is counted. It would be useful if I could set a tollerance term, like `early_stopping_tolerance=1e-3`.

From a quick survey, it looks like this is a common knob in other GBDT libraries (XGB, LGBM, CatBoost, scikit).

**Comment by jeremiedb** [comment]
@tecosaur For curiosity, since I've not been exposed to much cases where the out-of-sample metric don't end-up deteriorating, could you share some details about the context of this classification problem?
Ex:, number of bservations, features, class imbalance, etc? Also, is the model capacity large (max_depth, nbins)?
I'm asking as typically, GBT models overfit and exhibit gaps between training and evaluation data, such that out-of-sample end up having a clear deterioration rather than random oscillation. 
What was the `nrounds`?

**Comment by tecosaur** [comment]
Sure! Here's the situation

- Binary classification task with `EvoTreesClassifier`
- 61 features
- 340k rows/observations
- ~3% positive rate
- Max depth 6
- Bagging size 12
- nrounds 2000
- nbins 224
- eta 0.10

The rows are comprised of groups of ~12 (1-25) within which only one feature changes. Since I can't make the subsampling take this into account (this is the situation described in #287), I expect this leakage into the out-of-sample data causes the distributions to be very similar, and so unable to effectively assess when overfitting is happening.

I've tried to address this by using a group-aware splitting of the data into five folds, and then training five classifiers on 4/5 folds and evaluating the final model on the out-of-fold data (as you recommended).

**Comment by jeremiedb** [comment]
Thanks for the info. Have you seen bagging size had a substantial benefit? When having a small enough learning-rate leading to a large number of trees (a few hundred or more, like in you situation), I haven't observed yet a compelling benefit from this hyper-paramater. 
Had you performed hyper-search, notably with greater depth? I can be worth testsing with up to the 10-12 range. Maybe there were some represenattion capacity involved. 
Also, did you used `logloss` Regressor of the Regressor using softmax? (just for info, no direct implication I could think of).

I have no short-term solution regarding the grouping situation, but more group-wise capabilities is definitely on my radar. Just no short-term prioritization at the moment.

**Comment by tecosaur** [comment]
From my evaluations, bagging stabilises/improves the quality of the models built with very few positive data points. This was however tested with a lower depth earlier in the hyperparameter optimisation process.

The depth of 6 was the result of testing up to depth 12, and finding no benefit beyond 6.

Oh, and I've been using `:mlogloss`.

> I have no short-term solution regarding the grouping situation, but more group-wise capabilities is definitely on my radar. Just no short-term prioritization at the moment.

This is interesting to hear. Do you have much of an idea what form group-wise capabilities might take?

**Comment by singhharsh1708** [comment]
This is on main already, it landed in #331 as `early_stopping_tolerance`.

```julia
using EvoTrees, Random

Random.seed!(1)
x = rand(2_000, 4)
y = Float64.(x[:, 1] .> 0.5)

for tol in (0.0, 1e-3)
    config = EvoTreeRegressor(loss=:logloss, metric=:logloss, nrounds=400, eta=0.1,
                              early_stopping_rounds=10, early_stopping_tolerance=tol)
    m = EvoTrees.fit(config; x_train=x, y_train=y, x_eval=x, y_eval=y, verbosity=0)
    println("tol=$tol  nrounds_run=$(m.info[:nrounds])  best_iter=$(m.info[:logger][:best_iter])")
end
```

```
tol=0.0    nrounds_run=400  best_iter=400
tol=0.001  nrounds_run=78   best_iter=68
```

The check is `metric < best_metric - tol` in `src/callback.jl`, so it is an absolute tolerance, and it defaults to `0.0` so existing behaviour is unchanged. Is there a reason to keep this open?

**Comment by tecosaur** [comment]
I don't think so
